File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        maxarea=0

        stack=[] # index , height

        for i in range (len(heights)):

            start =i
            while stack and stack[-1][1] > heights[i]:

                currI, currH = stack.pop()

                maxarea = max((i-currI)*currH,maxarea)
                start= currI
            
            stack.append([start,heights[i]])
        
        while stack:
            currI, currH = stack.pop()

            maxarea = max((len(heights)-currI)*currH,maxarea)
        return maxarea

File: Data Structures & Algorithms/max-water-container/submission-0.java
class Solution {
    public int maxArea(int[] heights) {
        
        int maxVol = 0;

        int r = heights.length-1;
        int l =0 ;

        while( l<r)
        {
            int vol = (r-l)*(Math.min(heights[l],heights[r]));

            maxVol = Math.max(maxVol,vol);

            if(heights[l]<heights[r])
                l++;
            else
                r--;
        }
        return maxVol;
    }
}
